### PR TITLE
WIP: Refactor MDField memoization

### PR DESCRIPTION
**Pull request by jewatkins**
This changes the way MDField memoization is implemented in phalanx evaluators. Currently, the memoization flag has to be injected in every evaluator where the dev. wants to prevent recomputing. It's passed in through a parameter list during the construction of each evaluator. This process was a bit excessive and could lead to errors if the phalanx graph isn't followed explicitly.

This PR removes this method of initialization. Instead a class, PHAL::Setup, is created which gets passed into every phalanx evaluator during postRegistrationSetup. This allows us to bring in information from the app level directly into every evaluator. This PR also adds a map of MDField dependencies and a list of saved/unsaved fields (an example is attached [here](https://github.com/SNLComputation/Albany/files/2862432/fieldLists.txt) for the FO GIS unstructured test case).

By using the map of field dependencies, the dev. specifies which MDField will change (based on what happens outside of phalanx) and Phal::Setup can be used to update the list of saved/unsaved fields in the entire phalanx graph. For example, the solution always changes so a "unsaved" flag is passed into Phal::Setup and the list of saved/unsaved fields is updated based on what is evaluated from the solution.

Another feature is that Phal::Setup can be used to check whether all saved/unsaved fields match up with the MDFields in the field manager, thus ensuring that memoization is only used when the field dependency map is correct.

I'm still not sure what's the best way to pass the "unsaved" flag. Can we identify what gets changed from the input parameter list (e.g. check if ROL/LOCA is active and pass logic into the constructor/postRegistrationSetup of an evaluator)? Should we do proper memoization on the MDFields at the bottom of the trees (i.e. check whether the input is different from evaluated MDField)?

The PR is still WIP because I've only tested it with LandIce. I still need to expand it throughout Albany and test the other physics.

### CODE REVIEW

**Comment by mperego** [comment]
> @mperego There's a comment around line 913 of Albany_Application.cpp with your mark:
> 
> ```
>   // MPerego: Preforming post registration setup here to make sure that the
>   // discretization is already created, so that  derivative dimensions are known.
>   // Cannot do post registration right before the evaluate , as done for other
>   // field managers.  because memoizer hack is needed by Aeras.
>   // TODO, determine when it's best to perform post setup registration and fix
> // memoizer hack if needed.
> ```
> 
> Should we fix this now? I'm not sure why the response post registration setup could not happen right before the evaluation like the other field managers. Maybe the old memoizer was doing something weird.

@jewatkins Yes, if you know how to fix it go for it. It would take me some time to understand the logic of that comment.. but I trust you do :-)

**Comment by jewatkins** [comment]
> 
> 
> > @mperego There's a comment around line 913 of Albany_Application.cpp with your mark:
> > ```
> >   // MPerego: Preforming post registration setup here to make sure that the
> >   // discretization is already created, so that  derivative dimensions are known.
> >   // Cannot do post registration right before the evaluate , as done for other
> >   // field managers.  because memoizer hack is needed by Aeras.
> >   // TODO, determine when it's best to perform post setup registration and fix
> > // memoizer hack if needed.
> > ```
> > Should we fix this now? I'm not sure why the response post registration setup could not happen right before the evaluation like the other field managers. Maybe the old memoizer was doing something weird.
> 
> @jewatkins Yes, if you know how to fix it go for it. It would take me some time to understand the logic of that comment.. but I trust you do :-)

Okay I'll try moving it and see what happens.

**Comment by bartgol** [comment]
I haven't looked at the code, I just read your introduction at the top. This sounds like a great work.

Regarding your question on how to mark fields as unsaved: can we find a way to make a field "depend" on x,xdot,xdotdot,p (and maybe others) in the workset? If so, then we "only" need to put some logic in the workset from the app, marking x, xdot, xdotdot, or p as "old" or "new", and the evaluators can then decide what to do. Not sure how this last step can be done though. We could maybe come up with a time-stamp of sort for solution and parameters. The app could do the book keeping. Then during the workset setup, set these timestamps in the workset. All of this, assuming a PHX field can 'depend' on Thyra Vectors...

**Comment by jewatkins** [comment]
> Regarding your question on how to mark fields as unsaved: can we find a way to make a field "depend" on x,xdot,xdotdot,p (and maybe others) in the workset? If so, then we "only" need to put some logic in the workset from the app, marking x, xdot, xdotdot, or p as "old" or "new", and the evaluators can then decide what to do. Not sure how this last step can be done though. We could maybe come up with a time-stamp of sort for solution and parameters. The app could do the book keeping. Then during the workset setup, set these timestamps in the workset. All of this, assuming a PHX field can 'depend' on Thyra Vectors...

We may be able to hard code those structures in. The only thing I'm not sure about is which structures would need to be added and how the app knows they've been updated (i.e. app level book keeping). I don't think we need to worry about x, xdot, xdotdot because they change at every iteration, right? Thus, setting them as always unsaved seems okay for those. The ones that are difficult are the distributed parameters. Does Albany know at the app level whether those have been changed (or even if they going to be changing within the entire simulation)?

**Comment by bartgol** [comment]
I'm not sure that x, xdot, and xdotdot change at every call of evalModel. It's probably _generally_ true, but I don't really know how things work during the solution of the adjoint. I'm not sure if in that case the 'x' in the workset is the adjoint variable, or if it is the solution. In the former case, yes, x still changes, in the latter, it doesn't. @mperego you know more about how the adjoint is handled in Albany...

File: src/Aeras/evaluators/Aeras_ComputeBasisFunctions_Def.hpp [hunk 1]
```diff
@@ -143,6 +141,8 @@ postRegistrationSetup(typename Traits::SetupData d,
   DD3.setFieldData(ViewFactory::buildView(DD3.fieldTag(),ddims_));
 #endif
 
+  d.fill_field_dependencies(this->dependentFields(),this->evaluatedFields());
```
[the remark this answers is not in the document]
Comment (jewatkins) [reply src/Aeras/evaluators/Aeras_ComputeBasisFunctions_Def.hpp]: I decided to make this a requirement only when memoization is on.

**Comment by mperego** [comment]
> I'm not sure that x, xdot, and xdotdot change at every call of evalModel. It's probably _generally_ true, but I don't really know how things work during the solution of the adjoint. I'm not sure if in that case the 'x' in the workset is the adjoint variable, or if it is the solution. In the former case, yes, x still changes, in the latter, it doesn't. @mperego you know more about how the adjoint is handled in Albany...

I think that the 'x' in the workset is always the variable and it can change.

**Comment by mperego** [comment]
Thanks @jewatkins. This seems to me a clean way to handle memoization.
I think we could use observers to let Piro/Tempus/Loca/ROL mark as unsaved the fields that are being modified. What do you think? I can take care of the ROL/optimization part once we agree on this.

**Comment by bartgol** [comment]
@mperego How can we do that? We don't know _exactly_ when observers are called. Or, better, can we rely on observer to mark fields as unsaved without 'missing' any iteration?

**Comment by mperego** [comment]
In the case of ROL, we can call the observer before every evaluation of the model evaluator. Would that work?

**Comment by bartgol** [comment]
That's plenty. But are we going to be able to know if parameters have changed?

**Comment by mperego** [comment]
Yes, ROL has a function update that tells you whether the parameter has been updated.

**Comment by jewatkins** [comment]
I'm not sure what an observer does. Can you point me to an example? If it happens before the evaluation of the model evaluator, it may work.

**Comment by jewatkins** [comment]
Also, could we create a running list of all the fields that could change in LandIce (just to reinforce my understanding)? For example,

LOCA changes Glen's Law Homotopy Parameter - see input_fo_gis_unstructT.yaml test case

That should give us a nice list of tests to work through to make sure everything is working properly.

**Comment by mperego** [comment]
I was thinking at the observer as a C++ design pattern (e.g. https://cpppatterns.com/patterns/observer.html). 
Albany uses observers to have Piro/NOX print the solution to mesh after every nonlinear solve (see the `Albany_PiroObserverT.hpp` in the src folder). Basically, the class `PiroObserverT` implements a method `observeSolution` that prints the solution to mesh. The class `PiroObserverT` is passed to piro, and piro will call the function `observeSolution`, whenever it's time to print the mesh.

I was thinking of having another (or the same) observer class that implements a method, say `mark_field_unsaved(std::string fieldname)` that would mark a filed as unsaved. This method will be called by piro/ROL/NOX/LOCA right before evaluating the model evaluator, when needed.

**Comment by jewatkins** [comment]
Yeah that should work. The only issue I foresee is the fieldname. Phalanx uses a very specific identifier to differentiate between fields with the same name. We'll have to create some sort of mapping from PIRO/ROL/NOX/LOCA fieldnames to MDField names or identifiers. We would have that issue regardless of how we approach the problem though so I think the observers should work.

**Comment by rppawlo** [comment]
> Phalanx uses a very specific identifier to differentiate between fields with the same name.

you can actually take full control over the naming scheme, but you either have to write your own Tag or DataLayout. The DynamicLayout, for example can take any user supplied identifier string or defaults to an empty string.

**Comment by jewatkins** [comment]
Thanks for the info Roger. I suppose that's an option then but it'd be great if we can just create a mapping between the two so we don't have to create our own special layout for each parameter. I'm not sure how that would affect other functionality in Albany.

**Comment by ikalash** [comment]
Thank you for doing this @jewatkins , this is really nice! 

Regarding the Piro::TempusSolver interface: the relevant fields that can change here are x, xdot, and xdotdot.  If we are not solving a 2nd-order-in-time ODE, xdotdot will be null.  I do not envision a case where one of these fields would not be expected to change, so I think no memoization is required. 

Regarding in general parameters that can change: for quasistatic LCM problems, continuation is performed w.r.t. time using LOCA, which is stored as an MDField (please see the LCM::Time evaluator): https://github.com/SNLComputation/Albany/blob/master/src/LCM/evaluators/Time.hpp .  It looks like dt is a field as well.  @lxmota would be able to say more about how these fields are operated on.  I think if you have the plumbing working for the Piro::LOCASolver, it would be straightforward to memoize these fields.  For Tempus solves, time and dt are scalars, not MDFields.  For quasistatic Schwarz problems, we roll out our own continuation algorithm rather than using LOCA, but that is very specific.  I would suggest you focus on getting the memoization working for LandIce; then if other applications are interested in it, we can look at the LandIce examples.

**Comment by jewatkins** [comment]
I'm going to go ahead and push this as it doesn't break any tests. I'll add the observer for Piro input in another PR.

**Pull request closed without merging**